job_executor/worker/steps/dataset_pseudonymizer.py

```diff
@@ -1,15 +1,21 @@
 import logging
 from pathlib import Path
-from typing import Tuple, Union
+from typing import List, Optional, Tuple, Union
 
-import microdata_validator
+import pyarrow
+from pyarrow import dataset, compute, parquet
+
+import microdata_tools
+from microdata_tools.validation.model.metadata import UnitType
 
-from job_executor.exception import BuilderStepError
 from job_executor.adapter import pseudonym_service
+from job_executor.exception import BuilderStepError, UnregisteredUnitTypeError
 from job_executor.model import Metadata
 
 logger = logging.getLogger()
 
+VALID_UNIT_ID_TYPES = ["FNR"]
```

Comment:
We have a list of valid unit ID types in [microdata-tools](https://github.com/statisticsnorway/microdata-tools/blob/master/microdata_tools/validation/components/unit_id_types.py) so we could use it here.

Reply:
Yes I looked at it, and we use the type definition from it.  However we only want the ones that need to be psudomized past our guard statement?

Reply:
This is actually an interesting discussion. If we look at the table:
```py
UNIT_ID_TYPE_FOR_UNIT_TYPE = {
    "JOBB": "JOBBID_1",
    "KJORETOY": "KJORETOY_ID",
    "FAMILIE": "FNR",
    "FORETAK": "ORGNR",
    "HUSHOLDNING": "FNR",
    "KOMMUNE": None,
    "KURS": "KURSID",
    "PERSON": "FNR",
    "VIRKSOMHET": "ORGNR",
    "BK_HELSESTASJONSKONSULTASJON": "BK_STASJONS_BESOKS_ID",
}
```

Where a **unit_type** is the key
and a **unit_id_type** is the value

When we look at "KOMMUNE" it has no unit_id_type. This is because it does not require pseudonymization. (This reveals another flaw that is not yet handled: We should set BK_STATSJON_BESOK_ID to none.

The unit_id_type is ONLY used for pseudonymization as far as I know. It is an entirely internal term and definition.

Like my previous comment mentioned:
> Considering the code that generates the unit_id_type variable it can only be None or a valid unit_id_type 👍🏻 So there is no chance of an invalid unit_id_type at this point. This could be way clearer to readers of the code if we defined a UnitIdType enum in tools and used that as the type of the variable here ex.: Union[None, UnitIdType]


If you check your unittype with tools 3 things can happen:
- You get an UnregisteredUnitType error as your unitType does not exist in microdata
- You get a unit_id_type (which should always be pseudonymized)
- You get a None, indicating no pseudonymization needed

The best thing imho would be to make a UnitIdType enum in microdata-tools and use this for typing. This would make things much clearer.

The current branch now type annotes with UnitType:
```py
class UnitType(str, Enum):
    JOBB = "JOBB"
    KJORETOY = "KJORETOY"
    FAMILIE = "FAMILIE"
    FORETAK = "FORETAK"
    HUSHOLDNING = "HUSHOLDNING"
    KOMMUNE = "KOMMUNE"
    KURS = "KURS"
    PERSON = "PERSON"
    VIRKSOMHET = "VIRKSOMHET"
    BK_HELSESTASJONSKONSULTASJON = "BK_HELSESTASJONSKONSULTASJON"
```
But UnitType != UnitIdType. So the current type annotations are actually misleading. There is a wealth of domain specific terms and language at this exact point in the codebase 😱

Reply:
Oh so all unit ids that is not None should be pseudonymized?

I cant find any UnitIdType in tools. 
We currently use these tools methods to get the unit_type - key
```py
def _get_unit_types(
    metadata: Metadata,
) -> Tuple[Union[str, None], Union[str, None]]:
    return (
        metadata.get_identifier_key_type_name(),
        metadata.get_measure_key_type_name(),
    )
```
and if it exists we check for the value. 
```py
  identifier_unit_id_type = (
            None
            if identifier_unit_type is None
            else microdata_tools.get_unit_id_type_for_unit_type(
                identifier_unit_type
            )
        )
```

Then when we arrive at _pseudonumize_column we really only need to check for None since it should have been validated in the previous calls to tool. And thefore can remove the extra check we added and just use the original 

```py
if not unit_id_type:
  return None
```

Is this correct?

Also should we create UnitIdType? or is it hiding somewhere I can't find it?

Reply:
> Oh so all **UNIT ID TYPES** that is not None should be pseudonymized?

Yes 👍🏻 ( @pawbu back me up or disagree here please 😅 I am 99% sure after thinking about it)

> I cant find any UnitIdType in tools.

There is no enum definition of it yet. The only place it exists is in the unit_id_type_for_unit_type dictionary. This SHOULD be implemented as an enum though, for quality of life type hinting.

> and if it exists we check for the value.

Where the value would be a "unit_id_type" and the key we look up with would be a "unit_type" 👍🏻 

> Then when we arrive at _pseudonumize_column we really only need to check for None since it should have been validated in the previous calls to tool. And thefore can remove the extra check we added and just use the original

I would agree with this conclusion 👍🏻 Since the microdata-tools function results in a valid unit_id_type, none or a raised error, if no error was raised, we only really need to check if it is none 👍🏻 Again, this would be clearer if we had a enum in microdata-tools so it would be easier to type this when using tools 😅 That is my bad 🙏🏻

Reply:
@DanielElisenberg That is correct, I checked the code before this PR and every unit ID type that was returned from `microdata_validator.get_unit_id_type_for_unit_type()` function was pseudonymized.